File: js-lambda-build-image/boot/promise.ts
export class ProgrammablePromise<T> {
  private _resolve!: (value: T | PromiseLike<T>) => void;
  private _reject!: (reason?: any) => void;
  private _promise: Promise<T>;
  private _value!: T;

  constructor() {
    this._promise = new Promise<T>((resolve, reject) => {
      this._resolve = resolve;
      this._reject = reject;
    });
  }

  get promise() {
    return this._promise;
  }

  resolve(value: T | PromiseLike<T>) {
    this._resolve(value);
    this._value = value as T;
  }

  get reject() {
    return this._reject;
  }

  get value() {
    return this._value;
  }
}

File: python-lambda-build-image/boot/promise.py
"""Programmable Promise implementation for Python."""
import asyncio
from typing import Generic, TypeVar, Optional

T = TypeVar('T')

class ProgrammablePromise(Generic[T]):
    """A promise that can be resolved programmatically."""
    
    def __init__(self):
        self._future: asyncio.Future[T] = asyncio.Future()
        self._value: Optional[T] = None
    
    @property
    def promise(self) -> asyncio.Future[T]:
        """Get the underlying future."""
        return self._future
    
    def resolve(self, value: T) -> None:
        """Resolve the promise with a value."""
        if not self._future.done():
            self._future.set_result(value)
        self._value = value
    
    def reject(self, reason: Exception) -> None:
        """Reject the promise with an exception."""
        if not self._future.done():
            self._future.set_exception(reason)
    
    @property
    def value(self) -> Optional[T]:
        """Get the resolved value (if available)."""
        return self._value

